language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1204 @@
+/// It mimics execution on a single thread, with an call stack and an operand stack.
+/// The `Interpreter` receives a reference to a data store used by certain opcodes
+/// to do operations on data on chain and a `TransactionMetadata` which is also used to resolve
+/// specific opcodes.
+/// A `ModuleCache` is also provided to resolve external references to code.
+// REVIEW: abstract the data store better (maybe a single Trait for both data and event?)
+// The ModuleCache should be a Loader with a proper API.
+// Resolve where GasMeter should live.
+pub struct Interpreter<'alloc, 'txn, P>
+where
+    'alloc: 'txn,
+    P: ModuleCache<'alloc>,
+{
+    /// Operand stack, where MOVE `Value`s are stored for stack operations.
+    operand_stack: Stack,
+    /// The stack of active functions.
+    call_stack: CallStack<'txn>,
+    /// Gas metering to track cost of execution.
+    gas_meter: GasMeter,
+    /// Transaction data to resolve special bytecodes (e.g. GetTxnSequenceNumber, GetTxnPublicKey,
+    /// GetTxnSenderAddress, ...)
+    txn_data: TransactionMetadata,
+    /// List of events "fired" during the course of an execution.
+    // REVIEW: should this live outside the Interpreter?
+    event_data: Vec<ContractEvent>,
+    /// Data store
+    // REVIEW: maybe this and the event should go together as some kind of external context?
+    data_view: TransactionDataCache<'txn>,
+    /// Code cache, this is effectively the loader.
+    module_cache: P,
+    phantom: PhantomData<&'alloc ()>,
+}
+
+impl<'alloc, 'txn, P> Interpreter<'alloc, 'txn, P>
+where
+    'alloc: 'txn,
+    P: ModuleCache<'alloc>,
+{
+    /// Create a new instance of an `Interpreter` in the context of a transaction with a
+    /// given module cache (loader) and a data store.
+    // REVIEW: it's not clear the responsibilities between the Interpreter and the outside
+    // context are well defined. Obviously certain opcodes require a given context, but
+    // we may be doing better here...
+    pub fn new(
+        module_cache: P,
+        txn_data: TransactionMetadata,
+        data_view: TransactionDataCache<'txn>,
+    ) -> Self {
+        Interpreter {
+            operand_stack: Stack::new(),
+            call_stack: CallStack::new(),
+            gas_meter: GasMeter::new(txn_data.max_gas_amount()),
+            txn_data,
+            event_data: vec![],
+            data_view,
+            module_cache,
+            phantom: PhantomData,
+        }
+    }
+
+    //
+    // The functions below should be reviewed once we clean up the Loader and the
+    // transaction flow. It's not clear whether they are leaking internal of the Interpreter
+    // that would be better exposed via a more proper API.
+    //
+
+    /// Returns the module cache for this interpreter.
+    pub fn module_cache(&self) -> &P {
+        &self.module_cache
+    }
+
+    /// Disables metering of gas.
+    pub fn disable_metering(&mut self) {
+        self.gas_meter.disable_metering();
```

Comment:
I wonder if metering can be enabled/disabled statically? Doing this dynamically will require checking the bool flag & returning a result every time gas is consumed. Would it be better if we
1) Have a trait `GasMeter` and make the interpreter parametric over type `G: GasMeter`.
2) Implement the methods required by the trait for `()`, which does nothing internally. Ideally rust will be smart enough to see that all gas calls are no-ops when `()` is used as the meter.
3) Implement the methods for the normal gas meter.

We only disable metering for the epilogue & prologue so I assume the above would be possible? Thoughts?

Reply:
yeah those are possible solutions. This group of functions (as the comment above mentions) should be reviewed and possibly removed in favor of a different and better model.
The gas one in particular are tied to @tzakian's rework on gas metering which is changing things significantly and soon. So in that respect I am not sure what to say, except that is possible that `GasMeter` would be passed in and those functions would go away if so...
and I would hope they will be gone no matter what

Reply:
> Implement the methods required by the trait for (), which does nothing internally. Ideally rust will be smart enough to see that all gas calls are no-ops when () is used as the meter.

Yeah, Rust will do that